test/cpu.cpp

```diff
@@ -907,12 +919,232 @@ TEST_CASE("PLP", "[cpu][instruction]") {
+			/*addExpectedBusAccesses=*/noopAddBusAccesses,
+			/*setup=*/[&](CPU& cpu) {
+				cpu.e = 0;
+				cpu.A.forceStoreFull(val);
+			},
+			/*test=*/[&](CPU& cpu) {
+				REQUIRE(cpu.Y.load() == cpu.A.load());
+			}
+		);
+	}
+}
+
+TEST_CASE("TYA", "[cpu][instruction]") {
+	auto memoryAndAccumulatorAre8Bit = GENERATE(false, true);
+	auto val = GENERATE_COPY(take(1, random<Address>(0, (memoryAndAccumulatorAre8Bit ? 0xff : 0xffff) + 1)));
+		
+	DYNAMIC_SECTION((memoryAndAccumulatorAre8Bit ? 8 : 16) << "-bit") {
+		testInstruction(Opcode::TYA, AddressingMode::Implied,
+			/*addExpectedBusAccesses=*/noopAddBusAccesses,
+			/*setup=*/[&](CPU& cpu) {
+				cpu.e = 0;
+				cpu.Y.forceStoreFull(val);
+			},
+			/*test=*/[&](CPU& cpu) {
+				REQUIRE(cpu.A.load() == cpu.Y.load());
+			}
+		);
+	}
+}
+
+TEST_CASE("TXY", "[cpu][instruction]") {
+	auto indexRegistersAre8Bit = GENERATE(false, true);
+	auto val = GENERATE_COPY(take(1, random<Address>(0, (indexRegistersAre8Bit ? 0xff : 0xffff) + 1)));
+			
+	DYNAMIC_SECTION((indexRegistersAre8Bit ? 8 : 16) << "-bit") {
+		testInstruction(Opcode::TXY, AddressingMode::Implied,
+			/*addExpectedBusAccesses=*/noopAddBusAccesses,
+			/*setup=*/[&](CPU& cpu) {
+				cpu.e = 0;
+				cpu.X.forceStoreFull(val);
+			},
+			/*test=*/[&](CPU& cpu) {
+				REQUIRE(cpu.Y.load() == cpu.X.load());
+			}
+		);
+	}
+}
+
+TEST_CASE("TYX", "[cpu][instruction]") {
+	auto indexRegistersAre8Bit = GENERATE(false, true);
+	auto val = GENERATE_COPY(take(1, random<Address>(0, (indexRegistersAre8Bit ? 0xff : 0xffff) + 1)));
+			
+	DYNAMIC_SECTION((indexRegistersAre8Bit ? 8 : 16) << "-bit") {
+		testInstruction(Opcode::TYX, AddressingMode::Implied,
+			/*addExpectedBusAccesses=*/noopAddBusAccesses,
+			/*setup=*/[&](CPU& cpu) {
+				cpu.e = 0;
+				cpu.Y.forceStoreFull(val);
+			},
+			/*test=*/[&](CPU& cpu) {
+				REQUIRE(cpu.X.load() == cpu.Y.load());
+			}
+		);
+	}
+}
+
+TEST_CASE("TSX", "[cpu][instruction]") {
+	auto indexRegistersAre8Bit = GENERATE(false, true);
+	auto val = GENERATE_COPY(take(1, random<Address>(0, (indexRegistersAre8Bit ? 0xff : 0xffff) + 1)));
+			
+	DYNAMIC_SECTION((indexRegistersAre8Bit ? 8 : 16) << "-bit") {
+		testInstruction(Opcode::TSX, AddressingMode::Implied,
+			/*addExpectedBusAccesses=*/noopAddBusAccesses,
+			/*setup=*/[&](CPU& cpu) {
+				cpu.e = 0;
+				cpu.SP = val;
+			},
+			/*test=*/[&](CPU& cpu) {
+				REQUIRE(cpu.X.load() == cpu.SP);
+			}
+		);
+	}
+}
+
+TEST_CASE("TXS", "[cpu][instruction]") {
+	auto usingEmulatorMode = GENERATE(false, true);
+	auto val = GENERATE_COPY(take(1, random<Address>(0, (usingEmulatorMode ? 0xff : 0xffff) + 1)));
+			
+	DYNAMIC_SECTION((usingEmulatorMode ? 8 : 16) << "-bit") {
+		testInstruction(Opcode::TXS, AddressingMode::Implied,
+			/*addExpectedBusAccesses=*/noopAddBusAccesses,
+			/*setup=*/[&](CPU& cpu) {
+				cpu.e = usingEmulatorMode ? 1 : 0;
+				cpu.X.forceStoreFull(val);
+			},
+			/*test=*/[&](CPU& cpu) {
+				REQUIRE(cpu.SP == cpu.X.load());
```

Comment:
When emulation mode is enabled, the TXS instruction ensures that the high byte of the stack pointer is `0x01`.
```suggestion
				auto expectedVal = cpu.X.load() | (usingEmulatorMode ? 0x0100 : 0);
				REQUIRE(cpu.SP == expectedVal);
```